Dev old

francops1722 opened the pull request:
<!--
# nf-core/panoramaseq pull request

Many thanks for contributing to nf-core/panoramaseq!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/panoramaseq/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/panoramaseq/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, also make a PR on the nf-core/panoramaseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

francops1722 committed:
updated initialisation

francops1722 committed:
test modules updates

francops1722 committed:
tests updates

nvnieuwk commented on assets/samplesheet.csv:
```diff
@@ -1,3 +1,3 @@
-SAMPLE_PAIRED_END,/path/to/fastq/files/AEG588A1_S1_L002_R1_001.fastq.gz,/path/to/fastq/files/AEG588A1_S1_L002_R2_001.fastq.gz
-SAMPLE_SINGLE_END,/path/to/fastq/files/AEG588A4_S4_L003_R1_001.fastq.gz,
+sample,fastq_1,fastq_2,N_barcodes,barcode_file,sample_size,len_barcode,Nthresh,Ntriage
+S10,/user/gent/446/vsc44685/ScratchVO_dir/ST_transcriptomics_local/test/S10_UHRR_R1.fastq.gz,/user/gent/446/vsc44685/ScratchVO_dir/ST_transcriptomics_local/test/S10_UHRR_R2.fastq.gz,34500,/user/gent/446/vsc44685/ScratchVO_dir/panoramaseq/nf-core-panoramaseq/assets/barcodes_coords.csv,40000,36,9,1000
+S09,/user/gent/446/vsc44685/ScratchVO_dir/ST_transcriptomics_local/test/S09_UHRR_R1.fastq.gz,/user/gent/446/vsc44685/ScratchVO_dir/ST_transcriptomics_local/test/S09_UHRR_R2.fastq.gz,34500,/user/gent/446/vsc44685/ScratchVO_dir/panoramaseq/nf-core-panoramaseq/assets/barcodes_coords.csv,40000,36,9,1000
```
These local paths should be updated to urls so that this samplesheet works anywhere

nvnieuwk commented on main.nf:
```diff
@@ -1,73 +1,105 @@
+    // Otherwise, rely on params.star_genome_dir (backward compatibility)
+    //
+    if (params.fasta && params.star_gtf) {
+        ch_fasta = file(params.fasta, checkIfExists: true)
+        ch_gtf = file(params.star_gtf, checkIfExists: true)
```
This is a bit confusing since these aren't actually channels

nvnieuwk commented on main.nf:
```diff
@@ -1,73 +1,105 @@
+    // Run main PANORAMASEQ workflow
+    // This will handle all the steps defined in the PANORAMASEQ process including quality control, alignment, counting, etc.
+    // The PANORAMASEQ process is defined in the workflows/main.nf file
+    panorama_seq(
```
```suggestion
    PANORAMASEQ(
```

nvnieuwk commented on modules/local/Decoding/main.nf:
```diff
@@ -0,0 +1,68 @@
+    tag "$meta.id" 
+    label "gpu_process"
+    conda "${moduleDir}/environment.yml"
+    container 'docker://pytorch/pytorch:2.7.1-cuda11.8-cudnn9-runtime'
```
Preferably these containers should be created with [Seqera Containers](https://seqera.io/containers/) and both Docker and Singularity containers should be present. (Both can be created with Seqera Containers). Take a look at other pipelines to have an example on how this should be implemented

nvnieuwk commented on modules/local/Decoding/main.nf:
I won't repeat the last two comments for other modules, but these should also be applied there

nvnieuwk commented on modules/local/FeatCounts/main.nf:
The file path should be `modules/local/featurecounts/custom/main.nf` in this case

nvnieuwk commented on modules/local/FeatCounts/meta.yml:
```diff
@@ -0,0 +1,62 @@
+name: subread_featurecounts
```
```suggestion
name: featurecounts_custom
```

nvnieuwk commented on modules/local/cutadapt_adv_pipe/main.nf:
This module is missing a stub

francops1722 replied:
stub added

nvnieuwk commented on nextflow.config:
```diff
@@ -5,25 +5,46 @@
     Default config options for all compute environments
 ----------------------------------------------------------------------------------------
 */
+def scratch_dir =   System.getenv("VSC_SCRATCH_PROJECTS_BASE") ? "${System.getenv("VSC_SCRATCH_PROJECTS_BASE")}/${params.tier1_project}" : // Tier 1 scratch
+                    System.getenv("VSC_SCRATCH_VO_USER") ?: // VO scratch
+                    System.getenv("VSC_SCRATCH") // user scratch
```
This should not be in the pipeline config as it's system dependent. Please use the `vsc_ugent` profile instead to get the officially supported configuration for the UGhent HPC

francops1722 replied:
removed

nvnieuwk commented on nextflow.config:
```diff
@@ -5,25 +5,46 @@
+    additional_fasta           = null
     igenomes_base              = 's3://ngi-igenomes/igenomes/'
     igenomes_ignore            = false
+    star_genome_dir            = '/data/gent/vo/000/gvo00027/resources/STAR_index/Homo_sapiens/GRCh38.dna.primary_assembly.chrIS_spikes_45SwithGTF'   // <-- set your local STAR index directory here
+    star_gtf                   = '/data/gent/vo/000/gvo00027/resources/Ensembl_transcriptomes/Homo_sapiens/GRCh38/Homo_sapiens.GRCh38.109.chrIS_spikes_45S.gtf'   // <-- set your local GTF file here
```
These paths should not be hardcoded in the config

francops1722 replied:
done

nvnieuwk commented on nextflow.config:
```diff
@@ -140,6 +163,7 @@ profiles {
         podman.enabled          = false
         shifter.enabled         = false
         charliecloud.enabled    = false
+        apptainer.cacheDir = "$scratch_dir/singularity_cache"
```
```suggestion
```

francops1722 replied:
removed

nvnieuwk commented on subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf:
```diff
@@ -74,21 +74,17 @@ workflow PIPELINE_INITIALISATION {
-        .map {
-            meta, fastqs ->
-                return [ meta, fastqs.flatten() ]
+        .map { row ->
+            def meta = row[0]
+            def R1 = row[1]
+            def R2 = row[2]
```
```suggestion
        .map { meta, R1, R2 ->
```

You can directly assign variable names when a list has been given as input to the `.map` operators

nvnieuwk commented on utils/finalization.nf:
This subworkflow seems to be a copy of the [official nf-core subworkflow](https://github.com/nf-core/modules/tree/master/subworkflows/nf-core/utils_nfcore_pipeline), please just use that one instead :)

nvnieuwk commented on utils/panoramaseq_completion.nf:
This should be moved to the `subworkflows` folder

francops1722 replied:
done

nvnieuwk commented on workflows/ST_main/main.nf:
This can be removed

nvnieuwk commented on workflows/panoramaseq.nf:
This should be moved to `workflows/panoramaseq.nf`

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+workflow panorama_seq {
```
```suggestion
workflow PANORAMASSEQ {
```

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+    // 1. FastQC on raw input reads (valid_data stage)
+    fastqc_raw_input = valid_data.map { meta, reads ->
+        def new_meta = meta.clone()
+        new_meta.id = "raw_${meta.id}"
```
```suggestion
        def new_meta = meta + [id:"raw_${meta.id}"]
```
Using `.clone()` is a bad practice. This does not create a deep clone so the new object is still linked to the old object which can cause issues in a parallelized environment

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+        [new_meta, reads]
+    }
+    FASTQC(fastqc_raw_input)
+    ch_versions = ch_versions.mix(FASTQC.out.versions)
```
```suggestion
    ch_versions = ch_versions.mix(FASTQC.out.versions.first())
```
You only need one versions file per process, this can save some resources in very big runs especially. Can you apply this to each process?

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+
+    // 4. Combine both channels for downstream processing
+    // Use concat instead of mix to avoid multi-channel operator issues
+    umitools_input = SEQTK_SAMPLE.out.reads.concat(passthrough)
```
```suggestion
    umitools_input = SEQTK_SAMPLE.out.reads.mix(passthrough)
```
`.mix()` should work here. Especially since you don't use a multi-channel operator here. `.concat` waits for the `SEQTK_SAMPLE` process is fully done before add the files in `passthrough`, this creates unwanted bottlenecks

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+
+    // 6a. FastQC on reads after first trimming (cutadapt_results stage)
+    fastqc_cutadapt_input = CUTADAPT.out.reads.map { meta, reads ->
+        def new_meta = meta.clone()
+        new_meta.id = "cutadapt_${meta.id}"
```
```suggestion
        def new_meta = meta + [id: "cutadapt_${meta.id}"]
```

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+
+    // 8a. FastQC on reads after advanced trimming (cutadapt2_results stage)
+    fastqc_cutadapt2_input = CUTADAPT_ADV_PIPE.out.reads.map { meta, reads ->
+        def new_meta = meta.clone()
+        new_meta.id = "cutadapt2_${meta.id}"
```
```suggestion
        def new_meta = meta + [id:"cutadapt2_${meta.id}"]
```

nvnieuwk commented on workflows/panorama_seq/main.nf:
```diff
@@ -0,0 +1,276 @@
+    //     Uses the provided GTF annotation file
+    //     First update metadata to reflect single-end nature after R2-only alignment
+    star_bam_corrected_meta = STAR_ALIGN_LOCAL.out.bam.map { meta, bam ->
+        def new_meta = meta.clone()
+        new_meta.single_end = true  // Update to single_end since we only aligned R2
```
```suggestion
        def new_meta = meta + [single_end: true]  // Update to single_end since we only aligned R2
```

francops1722 committed:
Update main.nf

removing initialization info

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

replacing lowercase to uppercase

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/Decoding/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/Decoding/tests/nextflow.config

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/FeatCounts/meta.yml

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/UMI_count/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/anndata/check_h5ad/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update modules/local/anndata/make_h5ad/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update workflows/panorama_seq/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 commented on modules/local/SAMTOOLS_SORT_LOCAL/main.nf:
I made small modifications, basically I deleted the reference.fasta as input and  unnecessary outputs.

nvnieuwk replied:
You can use `[]` for as values for optional inputs, that way you don't have to change every module

francops1722 replied:
done! currently using the nf-core samtools sort module

francops1722 commented on modules/local/Decoding/environment.yml:
done

francops1722 commented on nextflow.config:
```diff
@@ -94,16 +119,14 @@ profiles {
+        singularity.enabled = true
+        singularity.autoMounts = true
+        singularity.cacheDir = "$scratch_dir/singularity_cache"
+        singularity.runOptions = '--nv'
```
removed

francops1722 commented on nextflow.config:
```diff
@@ -192,6 +216,8 @@ env {
     R_ENVIRON_USER   = "/.Renviron"
     JULIA_DEPOT_PATH = "/usr/local/share/julia"
+    APPTAINER_TMPDIR="$scratch_dir/.apptainer/tmp"
+    APPTAINER_CACHEDIR="$scratch_dir/.apptainer/cache"
```
removed

francops1722 committed:
many updates

francops1722 commented on modules/local/star_align/main.nf:
I think the issue was that the data is expected to be paired-end, but for the star alignment step it needs to be treated as single end. I don't remember if that was the only difference.

nvnieuwk replied:
The nf-core module does support single end fastqs, you just need to set `meta.single_end` to `true`

francops1722 committed:
Update pipeline configurations and integrate QUIK source code

- Updated test configs to use nf-core/test-datasets repository
- Updated anndata test modules to use nf-core test data
- Fixed MULTIQC input handling for UMITOOLS_EXTRACT logs
- Added QUIK source code as regular directory (not submodule)
- Updated custom config loading syntax per nf-core guidelines
- Added versions.yml snapshot test to workflow test
- Updated .gitignore to exclude build artifacts
- Removed decoding module (replaced by QUIK)

francops1722 committed:
update modules to match nf-core'

francops1722 committed:
solving failed lint

francops1722 committed:
force instal samtools index

francops1722 committed:
update readme

francops1722 committed:
readme updated

francops1722 committed:
update readme

francops1722 committed:
fix config gpus request

francops1722 committed:
handling empty h5ad files

francops1722 committed:
removing system dependant variables from configs

francops1722 committed:
delete quik source code

francops1722 committed:
remove system dependant config and update readme

francops1722 committed:
update readme no conda

francops1722 committed:
fix: address nf-core lint issues - add version snapshots to tests and update NF test matrix

francops1722 committed:
chore: stage nextflow.config changes

francops1722 committed:
fix: sync RO-Crate metadata with README.md description

francops1722 committed:
fix: sync template files with nf-core standards (v3.5.2)

francops1722 committed:
chore: update nf-core subworkflows to latest versions

francops1722 committed:
docs: update CHANGELOG for nf-core lint fixes

francops1722 reviewed (commented):
@nvnieuwk  i have run all tests sucesfully, updated the template and docs. Can you please reviewed this pull request so I can merge it with the dev branch?

francops1722 committed:
fix: correct UTILS_NFSCHEMA_PLUGIN workflow invocation with required 9 inputs

francops1722 committed:
style: fix code formatting - remove trailing whitespace, fix line endings, and correct indentation

- Remove trailing whitespace from all files
- Fix line endings (CRLF -> LF)
- Add missing final newlines
- Fix indentation to multiples of 4 spaces
- Addresses editorconfig-checker violations for nf-core CI

francops1722 committed:
chore: remove .nf-test/ directory from version control

This directory is already in .gitignore and should not be tracked

francops1722 committed:
style: format modules.json with prettier standards

francops1722 committed:
style: apply prettier formatting to all files

- Format README.md with consistent markdown styling
- Format YAML and JSON files with prettier standards
- Addresses all remaining prettier pre-commit violations

francops1722 committed:
fix: format custom profile loading with proper try/catch blocks

- Add try/catch blocks for custom config loading
- Uncomment pipeline-specific config loading
- Addresses nf-core lint nextflow_config requirements

francops1722 committed:
fix: consolidate try/catch blocks for custom profile loading

- Remove blank line between the two try/catch blocks
- Addresses nf-core lint nextflow_config parsing requirements

nvnieuwk commented on conf/base.config:
```diff
@@ -42,7 +41,7 @@ process {
     }
     withLabel:process_high {
-        cpus   = { 12    * task.attempt }
+        cpus   = { 6    * task.attempt }
```
```suggestion
        cpus   = { 12    * task.attempt }
```
Please don't change these defaults, some modules might break if you do so, you can manually set the amount of CPUs per process in the `conf/modules.config` file

nvnieuwk commented on modules/local/anndata/checkh5ad/main.nf:
Not entirely necessary for this PR but you should add tests to all local modules before the release

francops1722 replied:
by mistake i added tests to the git ignore file, its fixed now, and test modules are added.

nvnieuwk commented on modules/local/anndata/checkh5ad/main.nf:
```diff
@@ -0,0 +1,46 @@
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    echo "H5AD validation check completed successfully" > ${prefix}_h5ad_validation.log
+
```
You also need to create an empty h5ad file here with `touch <file>`

nvnieuwk commented on modules/local/cutadapt_adv_pipe/main.nf:
```diff
@@ -0,0 +1,38 @@
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    cutadapt -m 20 -O 20 -a "polyA=A{20}" -a "QUALITY=G{20}" -n 2 ${reads[1]} \
+    | cutadapt  -m 20 -O 3 --nextseq-trim=10  -a "r1adapter=A{18}AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC;min_overlap=3;max_error_rate=0.100000" - \
+    | cutadapt -m 20 -O 3 -a "r1polyA=A{18}" - \
+    | cutadapt -m 20 -O 20 -g "r1adapter=AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC;min_overlap=20" --discard-trimmed -o ${prefix}.trim2.R2.fastq.gz -
```
Is running the tool 4 times here the most efficient way to achieve this? Seems a bit excessive?

francops1722 replied:
we needed succesive trimming steps however in the future i think this step will be replaced. But for now I would prefer to keep it like this.

nvnieuwk replied:
Sounds good!

nvnieuwk commented on modules/local/cutadapt_panorama/main.nf:
This could probably be replaced by the nf-core module

francops1722 replied:
done replacing by the nf-core module

nvnieuwk commented on modules/local/samtoolssort/custom/main.nf:
This could also be replaced by the nf-core module

francops1722 replied:
the local module is actually not used, I will deleted non used modules to avoid confusion.

nvnieuwk commented on modules/local/staralign/custom/main.nf:
Same here

francops1722 replied:
done replacing by the nf-core module

nvnieuwk commented on subworkflows/local/prepare_genome/main.nf:
```diff
@@ -0,0 +1,81 @@
+    //-------------------------------------
+    // 1) Process FASTA file (decompress if needed)
+    //-------------------------------------
+    if (fasta.toString().endsWith('.gz')) {
```
```suggestion
    if (fasta.extension == 'gz') {
```

nvnieuwk commented on subworkflows/local/prepare_genome/main.nf:
```diff
@@ -0,0 +1,81 @@
+    //-------------------------------------
+    // 2) Process GTF file (decompress if needed)
+    //-------------------------------------
+    if (gtf.toString().endsWith('.gz')) {
```
```suggestion
    if (gtf.extension == 'gz') {
```

nvnieuwk commented on subworkflows/local/prepare_genome/main.nf:
```diff
@@ -0,0 +1,81 @@
+
+    if (additional_fasta) {
+        // Process additional FASTA (decompress if needed)
+        if (additional_fasta.toString().endsWith('.gz')) {
```
```suggestion
        if (additional_fasta.extension == 'gz') {
```

nvnieuwk commented on subworkflows/local/prepare_genome/main.nf:
```diff
@@ -0,0 +1,81 @@
+    //----------------------------------------------------
+    // 4) Generate STAR index
+    //----------------------------------------------------
+    STAR_GENOMEGENERATE(
+        ch_final_fasta.map { [ [:], it ] },
+        ch_gtf.map         { [ [:], it ] }
+    )
+    ch_versions = ch_versions.mix(STAR_GENOMEGENERATE.out.versions)
```
This has not yet been resolved

nvnieuwk commented on subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf:
```diff
@@ -74,21 +80,14 @@ workflow PIPELINE_INITIALISATION {
-            meta, fastqs ->
-                return [ meta, fastqs.flatten() ]
+        .map { meta, R1, R2 ->
+            if (!R2 || R2 == "" || R2 == "null") {
```
```suggestion
            if (!R2) {
```
This should be enough since Groovy considers `null` and `''` to be `false`

nvnieuwk commented on subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf:
```diff
@@ -74,21 +80,14 @@ workflow PIPELINE_INITIALISATION {
+        .map { meta, R1, R2 ->
+            if (!R2 || R2 == "" || R2 == "null") {
+                meta.single_end = true
+                return tuple(meta, [R1])
```
```suggestion
                def new_meta = meta + [single_end: true]
                return tuple(new_meta, [R1])
```
It's best practice to never overwrite the meta map directly but rather create a new one instead. This prevents unexpected race conditions where values get overwritten

nvnieuwk commented on subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf:
```diff
@@ -74,21 +80,14 @@ workflow PIPELINE_INITIALISATION {
+                return tuple(meta, [R1])
+            } else {
+                meta.single_end = false
+                return tuple(meta, [R1, R2])
```
```suggestion
                def new_meta = meta + [single_end: false]
                return tuple(new_meta, [R1, R2])
```

nvnieuwk commented on main.nf:
```diff
@@ -1,73 +1,98 @@
+        ch_star_index = PREPARE_GENOME.out.index
+        ch_gtf_file = PREPARE_GENOME.out.gtf
+
+    } else if (params.fasta && params.star_gtf && !params.star_genome_dir) {
```
Shouldn't `PREPARE_GENOME` also be run in this case?

francops1722 replied:
The prepare genome is a wrapper of genome generate that append extra fasta sequences (spike-ins) in addittion to the generation of the genome index. When no extra fasta files are provided, it only runs the genome generation directly without using prepare_genome. I have changed this know to use prepare genome in both cases, with additional fasta = null.

nvnieuwk replied:
Thanks, this makes sense!

nvnieuwk commented on nextflow.config:
```diff
@@ -286,3 +330,22 @@ validation {
+        }
+    }
+}
+
```
This seems very specific to our HPC and won't work on any other system. There already is a gpu label in the VSC_UGHENT profile which you can use. Please remove this from the pipeline as this config should be as generic as possible

https://github.com/nf-core/configs/blob/e521fbb6e2490db68ba5f529d3ec1fcf289e712f/conf/vsc_ugent.config#L46-L57

You can simply use `-profile vsc_ugent` with your pipeline to activate the config

nvnieuwk reviewed (commented):
Thanks for all the work! This is a massive improvement. There still are some minor issues but nothing major left. I'll be happy to approve once you've fixed those comments. Well done 🥳

francops1722 committed:
Update conf/base.config

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update subworkflows/local/prepare_genome/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update subworkflows/local/prepare_genome/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
Update subworkflows/local/utils_nfcore_panoramaseq_pipeline/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

francops1722 committed:
refactor: improve pipeline code quality and consistency

- Use PREPARE_GENOME subworkflow for both genome preparation paths (with/without additional_fasta)
- Fix file extension checking: use .extension property instead of .toString().endsWith()
- Improve meta map immutability: create new maps instead of mutating originals
- Add stub test data creation for ANNDATA_CHECKH5AD module
- Clean up nextflow.config: remove deprecated parameters, simplify profiles
- Update .gitignore to exclude test result directories
- Update modules.json with latest nf-core module versions

francops1722 committed:
feat: migrate from local STAR_ALIGN_LOCAL to nf-core STAR_ALIGN module

- Replace custom STAR_ALIGN_LOCAL with standardized nf-core STAR_ALIGN module
- Update input channel preparation to match nf-core module signature
- Add proper single_end flag handling in meta map
- Configure STAR_ALIGN with same alignment parameters as before
- Update version channel handling for topic-based emissions
- Use bam_sorted_aligned output channel for downstream processing
- Remove obsolete commented code from workflow header
- Add --readFilesCommand zcat to module config for .gz file support

Benefits:
- Better maintainability with standard nf-core module
- Automatic read group tags (ID, SM, PL, CN)
- More output options (junctions, unmapped reads, etc.)
- Easier updates through nf-core modules system

francops1722 committed:
feat: migrate from local CUTADAPT_PANORAMA to nf-core CUTADAPT module

- Replace custom CUTADAPT_PANORAMA with standardized nf-core CUTADAPT module
- Remove import alias since workflow always receives paired-end data at this stage
- Consolidate duplicate CUTADAPT configs in modules.config
- Add missing pipeline parameters: sample_size, star_gtf, star_genome_dir
- Configure STAR_ALIGN CPU limits for test profile (4 CPUs instead of 12)
- Remove topic-based version collection for STAR_ALIGN (auto-collected)

Benefits:
- Simpler codebase with standard nf-core module
- Better maintainability through nf-core updates
- Removes unused single-end filtering logic (never triggered in workflow)
- Consistent with other nf-core module usage

francops1722 committed:
chore: remove obsolete local modules after migration to nf-core

Removed modules that have been migrated to nf-core versions:
- modules/local/cutadapt_panorama -> nf-core/cutadapt
- modules/local/staralign -> nf-core/star/align
- modules/local/samtoolssort -> nf-core/samtools/sort

Added nf-core/star/align module files to track the newly installed module.

francops1722 committed:
chore: remove custom GPU configuration for nf-core compliance

nf-core pipelines only accept default settings, not custom GPU configurations.
Removed CUDA module loading and GPU cluster options.

francops1722 committed:
test: add nf-test module tests for ANNDATA_CHECKH5AD

Added comprehensive test suite for checkh5ad module including:
- Stub mode test
- Single sample tests (S09, S10)
- Multiple sample merge test
- Snapshot validation for output files

francops1722 committed:
test: add nf-test module tests for ANNDATA_MAKEH5AD and ANNDATA_MAKEH5ADSINGLE

Added comprehensive test suites for remaining anndata modules:
- ANNDATA_MAKEH5AD: tests for multi-sample H5AD creation
- ANNDATA_MAKEH5ADSINGLE: tests for single-sample H5AD creation
- Snapshot validation for output files

All anndata module tests were previously ignored by tests/ gitignore rule.

francops1722 committed:
test: add nf-test module tests for remaining local modules

Added comprehensive test suites and configs for:
- QUIK_BARCODE_CALLING: GPU-based barcode calling tests
- FEATURECOUNTS_CUSTOM: custom feature counting tests
- UMICOUNT: UMI counting tests

All local module tests were previously ignored by tests/ gitignore rule.
Module tests should be tracked separately from pipeline integration tests.

francops1722 committed:
fix: update .gitignore to only ignore top-level tests directory

Changed 'tests/' to '/tests/' to make the ignore rule root-relative.

This allows module tests (modules/local/*/tests/) to be tracked in git
while still ignoring the top-level tests/ directory for pipeline
integration tests.

Prevents future need to force-add module test files with -f flag.

nvnieuwk commented on nextflow.config:
```diff
@@ -128,14 +125,8 @@ profiles {
-        apptainer.enabled       = false
+        singularity.enabled = true
+        singularity.autoMounts = true
     }
```
Please undo this change, this is template code that should not be touched to stay compatible with all systems

francops1722 replied:
done

nvnieuwk commented on nextflow.config:
```diff
@@ -275,10 +256,10 @@ manifest {
             github: '',
-            contribution: ['author'], // List of contribution types ('author', 'maintainer' or 'contributor')
+            contribution: [], // List of contribution types ('author', 'maintainer' or 'contributor')
             orcid: ''
```
Can you fill these back in to get proper credit for your work?

francops1722 replied:
done

nvnieuwk commented on nextflow.config:
```diff
@@ -330,22 +311,8 @@ validation {
+//process {
+//    singularity.enabled = true
+//    singularity.runOptions = '--nv'
+//}
```
This can be removed now

francops1722 replied:
done

nvnieuwk reviewed (approved):
Some small comments, still some failing CI checks that need to be resolved, but feel free to merge once everything has been fixed

nvnieuwk commented:
@nf-core-bot fix linting

nf-core-bot committed:
[automated] Fix code linting

francops1722 committed:
test

francops1722 committed:
fix: resolve nf-core linting issues to meet pipeline standards

- Update .gitignore to exclude .venv/, venv/, __pycache__/, and lint output files
- Upgrade nf-core version from 3.2.1 to 3.5.2 in .nf-core.yml
- Add missing coords_csv parameter to nextflow.config
- Update custom profiles loading syntax to latest nf-core standard

This brings the pipeline to 322 tests passed with 0 failures.

Made-with: Cursor

francops1722 committed:
fix: resolve remaining nf-core linting failures

- Remove TODO comment from docs/output.md and add actual output documentation
- Remove /tests/ from .gitignore to allow test files to be tracked
- Add tests/default.nf.test to repository (was being ignored)

This fixes the 2 hard failures reported in CI linting:
1. pipeline_todos: TODO string in output.md
2. files_exist: File not found: tests/default.nf.test

All nf-core lint checks should now pass.

Made-with: Cursor

francops1722 committed:
fix: resolve pre-commit and linting CI failures

Pre-commit fixes:
- Fix prompt.json: use valid JSON syntax (quoted keys, no trailing comma)
- Add final newline to prompt.json
- Update .gitattributes to enforce LF line endings for all text files
- Apply Prettier formatting to modules.json and ro-crate-metadata.json

Linting fix:
- Change conf/modules.config: UMI_count -> UMICOUNT to match actual process name

These changes fix:
1. pre-commit prettier hook failures
2. editorconfig line ending/newline checks
3. nf-core modules_config lint test (withName must match existing process)

All CI checks should now pass.

Made-with: Cursor

francops1722 committed:
fix: apply prettier formatting to resolve pre-commit failures

Apply Prettier formatting according to .prettierrc.yml configuration:
- prompt.json: Use 4-space indentation (default tabWidth)
- modules.json: Use 4-space indentation and compact single-element arrays
- docs/output.md: Add blank line before bullet lists (markdown formatting)

All files now pass prettier checks. This fixes the pre-commit hook failures
where prettier was modifying files during CI runs.

Made-with: Cursor

francops1722 committed:
fix: restore .gitattributes to nf-core template

Replace custom .gitattributes with the exact nf-core template version to pass
the files_unchanged lint test. The custom LF line ending rules were redundant
as .editorconfig already enforces end_of_line=lf for all files.

Template contains only linguist overrides:
- Mark .config and .nf.test files as nextflow language
- Mark nf-core modules/subworkflows as generated code

This fixes the files_unchanged lint failure (321 -> 322 tests passed).

Made-with: Cursor

francops1722 committed:
fix: add missing final newlines for editorconfig compliance

Add final newlines to files that were missing them:
- .gitignore
- nextflow.config
- ro-crate-metadata.json

This fixes the editorconfig-checker pre-commit hook failures reporting
"Wrong line endings or no final newline" (8 errors found).

All files now comply with .editorconfig rules.

Made-with: Cursor

francops1722 committed:
fix: remove trailing whitespace for editorconfig compliance

Remove trailing whitespace from:
- conf/test.config (line 19)
- modules/local/featurecounts/custom/tests/nextflow.config (line 3)
- modules/local/quik/tests/main.nf.test (lines 35, 41, 45, 49)

This fixes the remaining editorconfig-checker pre-commit hook failures
reporting 6 errors. All files now comply with .editorconfig rules:
- end_of_line = lf ✅
- insert_final_newline = true ✅
- trim_trailing_whitespace = true ✅

Made-with: Cursor
